internal/agent/device/device.go

```diff
@@ -0,0 +1,162 @@
+package device
+
+import (
+	"context"
+	"time"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/device/config"
+	"github.com/flightctl/flightctl/internal/agent/device/status"
+	"github.com/flightctl/flightctl/internal/client"
+	"github.com/flightctl/flightctl/internal/tpm"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/flightctl/flightctl/pkg/executer"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/klog/v2"
+)
+
+// Agent is responsible for managing the configuration and status of the device.
+type Agent struct {
+	name                   string
+	device                 v1alpha1.Device
+	deviceStatus           v1alpha1.DeviceStatus
+	deviceStatusCollector  *status.Collector
+	managementClient       *client.Management
+	managementEndpoint     string
+	managementCertFilePath string
+	agentKeyFilePath       string
+	caCertFilePath         string
+	logPrefix              string
+	fetchSpecInterval      time.Duration
+	fetchStatusInterval    time.Duration
+	syncConfigInterval     time.Duration
+	configController       *config.Controller
+}
+
+// NewAgent creates a new device agent.
+func NewAgent(
+	name string,
+	fetchSpecInterval time.Duration,
+	fetchStatusInterval time.Duration,
+	syncConfigInterval time.Duration,
+	caCertFilePath string,
+	managementCertFilePath string,
+	agentKeyFilePath string,
+	managementEndpoint string,
+	tpm *tpm.TPM,
+	executor executer.Executer,
+	logPrefix string,
+	configController *config.Controller,
+) *Agent {
+	return &Agent{
+		name:                   name,
+		fetchSpecInterval:      fetchSpecInterval,
+		fetchStatusInterval:    fetchStatusInterval,
+		syncConfigInterval:     syncConfigInterval,
+		caCertFilePath:         caCertFilePath,
+		managementCertFilePath: managementCertFilePath,
+		agentKeyFilePath:       agentKeyFilePath,
+		managementEndpoint:     managementEndpoint,
+		logPrefix:              logPrefix,
+		configController:       configController,
+		device: v1alpha1.Device{
+			ApiVersion: "v1alpha1",
+			Kind:       "Device",
+			Status:     &v1alpha1.DeviceStatus{},
+			Metadata: v1alpha1.ObjectMeta{
+				Name: &name,
+			},
+		},
+		deviceStatusCollector: status.NewCollector(tpm, executor),
+	}
+}
+
+type AgentGetter interface {
+	Get(ctx context.Context) (*v1alpha1.Device, error)
+}
+
+// Run starts the device agent reconciliation loop.
+func (a *Agent) Run(ctx context.Context) error {
+	fetchSpecTicker, err := util.NewTickerWithJitter(a.fetchSpecInterval, 100*time.Millisecond, 50)
+	if err != nil {
+		return err
+	}
+	defer fetchSpecTicker.Stop()
+	fetchStatusTicker, err := util.NewTickerWithJitter(a.fetchStatusInterval, 100*time.Millisecond, 50)
+	if err != nil {
+		return err
+	}
+	defer fetchStatusTicker.Stop()
+	configSyncTicker, err := util.NewTickerWithJitter(a.syncConfigInterval, 100*time.Millisecond, 50)
+	if err != nil {
+		return err
+	}
+	defer configSyncTicker.Stop()
+
+	for {
```

Comment:
Let's review in the near future when / based on which events we sync config or update status. For example:
* We currently periodically poll for spec updates. Maybe in the future we add a server push of spec updates, too. When the spec updates, is there any reason to not immediately call configController.Sync()?
* Which other events require us to run config.Sync()? Does it need to run periodically?
* Which events require us to fetch status? And if status changes, would we immediately want to notify the service?